File: server/ml.py
import sys
import json
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
from sklearn.preprocessing import LabelEncoder
import os

# Suppress warnings
import warnings
warnings.filterwarnings('ignore')

# Map frontend values to dataset values where they differ
SOIL_TYPE_MAP = {
    'Silt': 'Silty',
    'Silty': 'Silty',
    'Clay': 'Clay',
    'Loamy': 'Loamy',
    'Sandy': 'Sandy',
    'Peaty': 'Peaty',
}

IRRIGATION_TYPE_MAP = {
    'Rainfed': 'Rain-fed',
    'Rain-fed': 'Rain-fed',
    'Drip': 'Drip',
    'Sprinkler': 'Sprinkler',
    'Flood': 'Flood',
    'Manual': 'Manual',
}

def main():
    try:
        input_data = json.loads(sys.argv[1])
        print(f"DEBUG RAW INPUT: {json.dumps(input_data)}", file=sys.stderr)

        # Load dataset
        dataset_path = os.path.join(os.getcwd(), 'attached_assets', 'agriculture_dataset_1772556603012.csv')
        df = pd.read_csv(dataset_path)
        print(f"DEBUG DATASET SIZE: {len(df)}", file=sys.stderr)

        # Use ALL features including categorical ones for better discrimination
        numerical_features = ['Farm_Area(acres)', 'Fertilizer_Used(tons)', 'Pesticide_Used(kg)', 'Water_Usage(cubic meters)']
        categorical_features = ['Soil_Type', 'Irrigation_Type', 'Season']
        target_crop = 'Crop_Type'

        # Encode categorical features using LabelEncoders
        label_encoders = {}
        for col in categorical_features:
            le = LabelEncoder()
            df[col + '_enc'] = le.fit_transform(df[col])
            label_encoders[col] = le

        encoded_cat_features = [col + '_enc' for col in categorical_features]
        all_features = numerical_features + encoded_cat_features

        X = df[all_features]
        y_crop = df[target_crop]

        X_train, X_test, y_train_crop, y_test_crop = train_test_split(X, y_crop, test_size=0.2, random_state=42)

        # Train Crop Model with more estimators for better accuracy
        crop_model = RandomForestClassifier(n_estimators=100, random_state=42)
        crop_model.fit(X_train, y_train_crop)

        # Calculate metrics for Crop Model
        crop_preds = crop_model.predict(X_test)
        accuracy = accuracy_score(y_test_crop, crop_preds)
        precision = precision_score(y_test_crop, crop_preds, average='weighted', zero_division=0)
        recall = recall_score(y_test_crop, crop_preds, average='weighted', zero_division=0)
        f1 = f1_score(y_test_crop, crop_preds, average='weighted', zero_division=0)

        # Map frontend categorical values to dataset values
        raw_soil = input_data.get('soilType', 'Loamy')
        raw_irrigation = input_data.get('irrigationType', 'Drip')
        raw_season = input_data.get('season', 'Kharif')

        mapped_soil = SOIL_TYPE_MAP.get(raw_soil, raw_soil)
        mapped_irrigation = IRRIGATION_TYPE_MAP.get(raw_irrigation, raw_irrigation)
        mapped_season = raw_season  # Season values match between frontend and dataset

        print(f"DEBUG MAPPED CATEGORICALS: soil={mapped_soil}, irrigation={mapped_irrigation}, season={mapped_season}", file=sys.stderr)

        # Encode user categorical inputs using the fitted encoders
        # Handle unknown categories gracefully by defaulting to the most common class
        def safe_encode(encoder, value, col_name):
            if value in encoder.classes_:
                return encoder.transform([value])[0]
            else:
                print(f"WARNING: Unknown {col_name} value '{value}', known values: {list(encoder.classes_)}", file=sys.stderr)
                # Default to first class (most common after sorting)
                return 0

        soil_enc = safe_encode(label_encoders['Soil_Type'], mapped_soil, 'Soil_Type')
        irrigation_enc = safe_encode(label_encoders['Irrigation_Type'], mapped_irrigation, 'Irrigation_Type')
        season_enc = safe_encode(label_encoders['Season'], mapped_season, 'Season')

        user_df = pd.DataFrame([{
            'Farm_Area(acres)': float(input_data.get('farmArea', 100)),
            'Fertilizer_Used(tons)': float(input_data.get('fertilizerUsed', 5)),
            'Pesticide_Used(kg)': float(input_data.get('pesticideUsed', 2)),
            'Water_Usage(cubic meters)': float(input_data.get('waterUsage', 50000)),
            'Soil_Type_enc': soil_enc,
            'Irrigation_Type_enc': irrigation_enc,
            'Season_enc': season_enc,
        }])

        print(f"DEBUG ENCODED INPUT: {user_df.to_dict(orient='records')}", file=sys.stderr)

        pred_crop = crop_model.predict(user_df)[0]
        # Probability/Confidence
        probs = crop_model.predict_proba(user_df)[0]
        confidence = float(max(probs))

        print(f"DEBUG PREDICTION: crop={pred_crop}, confidence={confidence:.4f}", file=sys.stderr)
        print(f"DEBUG PROBABILITIES: {dict(zip(crop_model.classes_, [round(float(p), 4) for p in probs]))}", file=sys.stderr)

        # Mock disease prediction since it's not in dataset
        diseases = {
            'Cotton': 'Boll Rot',
            'Carrot': 'Leaf Blight',
            'Sugarcane': 'Red Rot',
            'Tomato': 'Late Blight',
            'Soybean': 'Rust',
            'Rice': 'Blast',
            'Maize': 'Stalk Rot',
            'Wheat': 'Powdery Mildew',
            'Barley': 'Net Blotch',
            'Potato': 'Early Blight'
        }
        pred_disease = diseases.get(pred_crop, 'None Detected')

        result = {
            "recommendedCrop": pred_crop,
            "diseasePrediction": pred_disease,
            "confidenceScore": confidence,
            "metrics": {
                "accuracy": accuracy,
                "precision": precision,
                "recall": recall,
                "f1Score": f1
            }
        }
        print(json.dumps(result))

    except Exception as e:
        import traceback
        print(f"ML ERROR: {str(e)}", file=sys.stderr)
        print(traceback.format_exc(), file=sys.stderr)
        sys.exit(1)

if __name__ == "__main__":
    main()

File: server/ml.ts
import fs from "fs";
import path from "path";

// ── Category Encoding Maps ──────────────────────────────────────────────
// Maps frontend values → numeric codes. Also handles dataset value variants.

const SOIL_TYPE_MAP: Record<string, number> = {
  Clay: 0,
  Loamy: 1,
  Peaty: 2,
  Sandy: 3,
  Silt: 4,
  Silty: 4, // dataset uses "Silty", frontend sends "Silt"
};

const IRRIGATION_TYPE_MAP: Record<string, number> = {
  Drip: 0,
  Flood: 1,
  Manual: 2,
  "Rain-fed": 3,
  Rainfed: 3, // frontend sends "Rainfed", dataset uses "Rain-fed"
  Sprinkler: 4,
};

const SEASON_MAP: Record<string, number> = {
  Kharif: 0,
  Rabi: 1,
  Zaid: 2,
};

// Disease map per crop
const DISEASE_MAP: Record<string, string> = {
  Cotton: "Boll Rot",
  Carrot: "Leaf Blight",
  Sugarcane: "Red Rot",
  Tomato: "Late Blight",
  Soybean: "Rust",
  Rice: "Blast",
  Maize: "Stalk Rot",
  Wheat: "Powdery Mildew",
  Barley: "Net Blotch",
  Potato: "Early Blight",
};

// ── Types ───────────────────────────────────────────────────────────────

interface DataRow {
  cropType: string;
  features: number[]; // [farmArea, fertilizer, pesticide, waterUsage, soilEnc, irrigationEnc, seasonEnc]
}

interface PredictionInput {
  farmArea?: number;
  fertilizerUsed?: number;
  pesticideUsed?: number;
  waterUsage?: number;
  soilType?: string;
  irrigationType?: string;
  season?: string;
}

interface PredictionResult {
  recommendedCrop: string;
  diseasePrediction: string;
  confidenceScore: number;
  metrics: {
    accuracy: number;
    precision: number;
    recall: number;
    f1Score: number;
  };
}

// ── Dataset Loading ─────────────────────────────────────────────────────

let cachedDataset: DataRow[] | null = null;
let cachedMetrics: { accuracy: number; precision: number; recall: number; f1Score: number } | null = null;

// Feature min/max for normalization
let featureMin: number[] = [];
let featureMax: number[] = [];

function loadDataset(): DataRow[] {
  if (cachedDataset) return cachedDataset;

  const csvPath = path.join(process.cwd(), "attached_assets", "agriculture_dataset_1772556603012.csv");
  console.log("ML: Loading dataset from", csvPath);

  if (!fs.existsSync(csvPath)) {
    throw new Error(`Dataset not found at ${csvPath}`);
  }

  const raw = fs.readFileSync(csvPath, "utf-8");
  const lines = raw.trim().split(/\r?\n/);

  if (lines.length < 2) {
    throw new Error("Dataset is empty or has no data rows");
  }

  // Parse header
  // Farm_ID,Crop_Type,Farm_Area(acres),Irrigation_Type,Fertilizer_Used(tons),Pesticide_Used(kg),Yield(tons),Soil_Type,Season,Water_Usage(cubic meters)
  const rows: DataRow[] = [];

  for (let i = 1; i < lines.length; i++) {
    const cols = lines[i].split(",");
    if (cols.length < 10) continue;

    const cropType = cols[1].trim();
    const farmArea = parseFloat(cols[2]);
    const irrigationType = cols[3].trim();
    const fertilizer = parseFloat(cols[4]);
    const pesticide = parseFloat(cols[5]);
    // cols[6] = Yield (not used as feature)
    const soilType = cols[7].trim();
    const season = cols[8].trim();
    const waterUsage = parseFloat(cols[9]);

    const soilEnc = SOIL_TYPE_MAP[soilType];
    const irrigationEnc = IRRIGATION_TYPE_MAP[irrigationType];
    const seasonEnc = SEASON_MAP[season];

    if (soilEnc === undefined || irrigationEnc === undefined || seasonEnc === undefined) {
      console.warn(`ML: Skipping row ${i} - unknown category: soil=${soilType}, irrigation=${irrigationType}, season=${season}`);
      continue;
    }

    if (isNaN(farmArea) || isNaN(fertilizer) || isNaN(pesticide) || isNaN(waterUsage)) {
      console.warn(`ML: Skipping row ${i} - invalid numeric values`);
      continue;
    }

    rows.push({
      cropType,
      features: [farmArea, fertilizer, pesticide, waterUsage, soilEnc, irrigationEnc, seasonEnc],
    });
  }

  console.log(`ML: Dataset loaded successfully. ${rows.length} valid rows.`);

  if (rows.length === 0) {
    throw new Error("No valid data rows after parsing");
  }

  // Compute feature min/max for normalization
  const numFeatures = rows[0].features.length;
  featureMin = new Array(numFeatures).fill(Infinity);
  featureMax = new Array(numFeatures).fill(-Infinity);

  for (const row of rows) {
    for (let j = 0; j < numFeatures; j++) {
      if (row.features[j] < featureMin[j]) featureMin[j] = row.features[j];
      if (row.features[j] > featureMax[j]) featureMax[j] = row.features[j];
    }
  }

  cachedDataset = rows;

  // Compute metrics via leave-one-out cross-validation
  cachedMetrics = computeMetrics(rows);
  console.log("ML: Model metrics:", cachedMetrics);

  return rows;
}

// ── Normalization ───────────────────────────────────────────────────────

function normalize(features: number[]): number[] {
  return features.map((v, i) => {
    const range = featureMax[i] - featureMin[i];
    if (range === 0) return 0;
    return (v - featureMin[i]) / range;
  });
}

// ── KNN Prediction ──────────────────────────────────────────────────────

const K = 5;

function euclideanDistance(a: number[], b: number[]): number {
  let sum = 0;
  for (let i = 0; i < a.length; i++) {
    const diff = a[i] - b[i];
    sum += diff * diff;
  }
  return Math.sqrt(sum);
}

function predictKNN(
  dataset: DataRow[],
  inputFeatures: number[]
): { crop: string; confidence: number; probabilities: Record<string, number> } {
  const normalizedInput = normalize(inputFeatures);

  // Calculate distances to all data points
  const distances: { distance: number; crop: string }[] = dataset.map((row) => ({
    distance: euclideanDistance(normalize(row.features), normalizedInput),
    crop: row.cropType,
  }));

  // Sort by distance ascending
  distances.sort((a, b) => a.distance - b.distance);

  // Take K nearest neighbors with distance weighting
  const kNearest = distances.slice(0, K);

  // Weighted voting: closer neighbors have more influence
  const cropWeights: Record<string, number> = {};
  let totalWeight = 0;

  for (const neighbor of kNearest) {
    // Inverse distance weighting (add small epsilon to avoid division by zero)
    const weight = 1 / (neighbor.distance + 0.0001);
    cropWeights[neighbor.crop] = (cropWeights[neighbor.crop] || 0) + weight;
    totalWeight += weight;
  }

  // Find the crop with highest weight
  let bestCrop = "";
  let bestWeight = -1;
  for (const [crop, weight] of Object.entries(cropWeights)) {
    if (weight > bestWeight) {
      bestWeight = weight;
      bestCrop = crop;
    }
  }

  // Calculate probabilities
  const probabilities: Record<string, number> = {};
  for (const [crop, weight] of Object.entries(cropWeights)) {
    probabilities[crop] = weight / totalWeight;
  }

  const confidence = bestWeight / totalWeight;

  return { crop: bestCrop, confidence, probabilities };
}

// ── Metrics (Leave-One-Out Cross-Validation) ────────────────────────────

function computeMetrics(dataset: DataRow[]): {
  accuracy: number;
  precision: number;
  recall: number;
  f1Score: number;
} {
  let correct = 0;
  const allCrops = [...new Set(dataset.map((r) => r.cropType))];
  const perClassTP: Record<string, number> = {};
  const perClassFP: Record<string, number> = {};
  const perClassFN: Record<string, number> = {};

  for (const crop of allCrops) {
    perClassTP[crop] = 0;
    perClassFP[crop] = 0;
    perClassFN[crop] = 0;
  }

  for (let i = 0; i < dataset.length; i++) {
    // Leave one out
    const testRow = dataset[i];
    const trainSet = dataset.filter((_, idx) => idx !== i);

    // Temporarily use trainSet for normalization
    const tmpMin = new Array(testRow.features.length).fill(Infinity);
    const tmpMax = new Array(testRow.features.length).fill(-Infinity);
    for (const row of trainSet) {
      for (let j = 0; j < row.features.length; j++) {
        if (row.features[j] < tmpMin[j]) tmpMin[j] = row.features[j];
        if (row.features[j] > tmpMax[j]) tmpMax[j] = row.features[j];
      }
    }

    const normTest = testRow.features.map((v, j) => {
      const range = tmpMax[j] - tmpMin[j];
      return range === 0 ? 0 : (v - tmpMin[j]) / range;
    });

    const dists: { distance: number; crop: string }[] = trainSet.map((row) => ({
      distance: euclideanDistance(
        row.features.map((v, j) => {
          const range = tmpMax[j] - tmpMin[j];
          return range === 0 ? 0 : (v - tmpMin[j]) / range;
        }),
        normTest
      ),
      crop: row.cropType,
    }));

    dists.sort((a, b) => a.distance - b.distance);
    const knn = dists.slice(0, K);

    // Simple majority vote for metrics
    const votes: Record<string, number> = {};
    for (const n of knn) {
      votes[n.crop] = (votes[n.crop] || 0) + 1;
    }
    let predicted = "";
    let maxVotes = 0;
    for (const [crop, count] of Object.entries(votes)) {
      if (count > maxVotes) {
        maxVotes = count;
        predicted = crop;
      }
    }

    const actual = testRow.cropType;
    if (predicted === actual) {
      correct++;
      perClassTP[actual]++;
    } else {
      perClassFP[predicted] = (perClassFP[predicted] || 0) + 1;
      perClassFN[actual] = (perClassFN[actual] || 0) + 1;
    }
  }

  const accuracy = correct / dataset.length;

  // Weighted precision, recall, F1
  let weightedPrecision = 0;
  let weightedRecall = 0;
  let weightedF1 = 0;
  const totalSamples = dataset.length;

  for (const crop of allCrops) {
    const classCount = dataset.filter((r) => r.cropType === crop).length;
    const tp = perClassTP[crop];
    const fp = perClassFP[crop];
    const fn = perClassFN[crop];

    const precision = tp + fp > 0 ? tp / (tp + fp) : 0;
    const recall = tp + fn > 0 ? tp / (tp + fn) : 0;
    const f1 = precision + recall > 0 ? (2 * precision * recall) / (precision + recall) : 0;

    const weight = classCount / totalSamples;
    weightedPrecision += precision * weight;
    weightedRecall += recall * weight;
    weightedF1 += f1 * weight;
  }

  return {
    accuracy: Math.round(accuracy * 100) / 100,
    precision: Math.round(weightedPrecision * 100) / 100,
    recall: Math.round(weightedRecall * 100) / 100,
    f1Score: Math.round(weightedF1 * 100) / 100,
  };
}

// ── Public API ──────────────────────────────────────────────────────────

export function predict(input: PredictionInput): PredictionResult {
  console.log("ML: RAW INPUT:", JSON.stringify(input));

  // Validate required fields exist
  if (!input.soilType) throw new Error("Missing required field: soilType");
  if (!input.irrigationType) throw new Error("Missing required field: irrigationType");
  if (!input.season) throw new Error("Missing required field: season");
  if (input.farmArea === undefined || input.farmArea === null) throw new Error("Missing required field: farmArea");
  if (input.fertilizerUsed === undefined || input.fertilizerUsed === null) throw new Error("Missing required field: fertilizerUsed");
  if (input.pesticideUsed === undefined || input.pesticideUsed === null) throw new Error("Missing required field: pesticideUsed");
  if (input.waterUsage === undefined || input.waterUsage === null) throw new Error("Missing required field: waterUsage");

  // Validate and encode categorical inputs
  const soilEnc = SOIL_TYPE_MAP[input.soilType];
  if (soilEnc === undefined) {
    throw new Error(`Invalid soilType: "${input.soilType}". Valid: ${Object.keys(SOIL_TYPE_MAP).join(", ")}`);
  }

  const irrigationEnc = IRRIGATION_TYPE_MAP[input.irrigationType];
  if (irrigationEnc === undefined) {
    throw new Error(`Invalid irrigationType: "${input.irrigationType}". Valid: ${Object.keys(IRRIGATION_TYPE_MAP).join(", ")}`);
  }

  const seasonEnc = SEASON_MAP[input.season];
  if (seasonEnc === undefined) {
    throw new Error(`Invalid season: "${input.season}". Valid: ${Object.keys(SEASON_MAP).join(", ")}`);
  }

  // Validate numerics
  const farmArea = Number(input.farmArea);
  const fertilizer = Number(input.fertilizerUsed);
  const pesticide = Number(input.pesticideUsed);
  const waterUsage = Number(input.waterUsage);

  if (isNaN(farmArea)) throw new Error(`Invalid farmArea: ${input.farmArea}`);
  if (isNaN(fertilizer)) throw new Error(`Invalid fertilizerUsed: ${input.fertilizerUsed}`);
  if (isNaN(pesticide)) throw new Error(`Invalid pesticideUsed: ${input.pesticideUsed}`);
  if (isNaN(waterUsage)) throw new Error(`Invalid waterUsage: ${input.waterUsage}`);

  const features = [farmArea, fertilizer, pesticide, waterUsage, soilEnc, irrigationEnc, seasonEnc];
  console.log("ML: ENCODED FEATURES:", features);

  // Load dataset (cached after first call)
  const dataset = loadDataset();
  console.log("ML: DATASET SIZE:", dataset.length);

  // Run KNN prediction
  const { crop, confidence, probabilities } = predictKNN(dataset, features);

  console.log("ML: PREDICTION:", crop, "CONFIDENCE:", confidence.toFixed(4));
  console.log("ML: PROBABILITIES:", probabilities);

  const diseasePrediction = DISEASE_MAP[crop] || "None Detected";

  return {
    recommendedCrop: crop,
    diseasePrediction,
    confidenceScore: Math.round(confidence * 100) / 100,
    metrics: cachedMetrics || { accuracy: 0, precision: 0, recall: 0, f1Score: 0 },
  };
}
